ledger/nomos-ledger/src/mantle/sdp/rewards.rs

```diff
@@ -275,8 +262,209 @@ impl Rewards for DaRewards {
+                        expected: session_state.session_n,
+                        got: proof.session,
+                    });
+                }
+
+                let num_declarations = session_state.declarations.size() as u64;
+                if num_declarations < settings.minimum_network_size.get() {
+                    return Err(Error::MinimumNetworkSizeNotSatisfied {
+                        num_declarations,
+                        minimum_network_size: settings.minimum_network_size,
+                    });
+                }
+
+                let proof = nomos_blend_message::reward::ActivityProof::try_from(proof)
+                    .map_err(|_| Error::InvalidProofType)?;
+
+                // TODO: Validate PoQ and PoSel by holding all epoch infos that
+                // spanned across the `session_state.session_n` session.
+
+                let hamming_distance = proof.token().hamming_distance(
+                    settings.expected_token_count_byte_len(num_declarations),
+                    *next_session_randomness,
+                );
+
+                Ok(Self::Initialized {
+                    session_state: session_state.clone(),
+                    next_session_randomness: *next_session_randomness,
+                    submitted_proofs: submitted_proofs.insert(provider_id, hamming_distance),
+                    min_hamming_distance: min(*min_hamming_distance, hamming_distance),
+                    settings: settings.clone(),
+                })
+            }
+        }
+    }
+
+    fn update_session(
+        &self,
+        last_active: &SessionState,
+        next_active_session_epoch_nonce: &Fr,
+        _config: &ServiceParameters,
+    ) -> (Self, HashMap<ProviderId, RewardAmount>) {
+        match self {
+            Self::Uninitialized(settings) => (
+                Self::Initialized {
+                    session_state: last_active.clone(),
+                    next_session_randomness: SessionRandomness::new(
+                        last_active.session_n + 1,
+                        next_active_session_epoch_nonce,
+                    ),
+                    submitted_proofs: HashTrieMapSync::new_sync(),
+                    min_hamming_distance: u64::MAX,
+                    settings: settings.clone(),
+                },
+                HashMap::new(),
+            ),
+            Self::Initialized {
+                submitted_proofs,
+                min_hamming_distance,
+                settings,
+                ..
+            } => {
+                // TODO: Calculate base rewards when session_income is added to config
+                // For now using placeholder value of 0
+                let session_income = 0;
+
+                // Identify premium providers with the minimum Hamming distance
+                let premium_providers = submitted_proofs
+                    .iter()
+                    .filter_map(|(&id, &hamming_distance)| {
```

Comment:
Why not keeping track directly of the providers with the smallest distance at the time when we evaluate a new active message instead?